Push notifications support via UnifiedPush

PeterCxy opened the pull request:
Fixes #793.

This is an implementation for push notifications based on UnifiedPush for Tusky. No push gateway (other than UP itself) is needed, since UnifiedPush is simple enough such that it can act as a catch-all endpoint for WebPush messages. When a UnifiedPush distributor is present on-device, we will by default register Tusky as a receiver; if no UnifiedPush distributor is available, then pull notifications are used as a fallback mechanism. 

Because WebPush messages are encrypted, and Mastodon does not send the keys and IV needed for decryption in the request body, for now the push handler simply acts as a trigger for the pre-existing NotificationWorker which is also used for pull notifications. Nevertheless, I have implemented proper key generation and storage, just in case we would like to implement full decryption support in the future when Mastodon upgrades to the latest WebPush encryption scheme that includes all information in the request body.

For users with existing accounts, push notifications will not be enabled until all of the accounts have been re-logged in to grant the new `push` OAuth scope. A small prompt will be shown (until dismissed) as a `Snackbar` to explain to the user about this, and an option is added in Account Preferences to facilitate re-login without deleting local drafts and cache.

connyduck commented:
This is awesome 😍 

Regarding the migration of old accounts - prompting the user to relogin is probably the best idea. Be careful not to delete the users local settings & drafts.

Please add no new code with Java & Rx, we are trying to replace it all with Kotlin & Coroutines right now 😅 
Instead of adding to NotificationHelper.java I would add a new file PushNotificationHelper.kt.

Mek101 commented:
Following this

karmanyaahm commented:
> Because WebPush messages are encrypted, and the UnifiedPush specification does not handle binary messages yet (but will eventually), for now the push handler simply acts as a trigger for the pre-existing NotificationWorker which is also used for pull notifications.

Just an update @PeterCxy: v2.0 of UP android-connector now handles binary messages properly

PeterCxy commented:
@karmanyaahm Thank you, I've been waiting for their binary message support for a while. I'll see if I can make use of that to do direct notification delivery later (after I finish my work for this semester).

connyduck commented:
I would really really like to have this feature in Tusky - will you continue working on it @PeterCxy? Is there anything I can help you with?

PeterCxy commented:
@connyduck I will be working on it -- hopefully I'll be back in the next few weeks or so, if everything goes well :)

PeterCxy commented:
I've rebased the changes to the latest develop branch + latest UnifiedPush connector 2.0. Currently looking into potentially implementing WebPush decryption (and re-registering the push notification types after changing it in settings)

karmanyaahm commented:
I don't know if you already have this, but this was posted to the UP chat a little while back by @emersion. It might be a good reference https://git.sr.ht/~emersion/PushGardenTestApp/tree/master/item/app/src/main/java/com/example/pushgardentestapp/WebPush.kt

PeterCxy commented:
@karmanyaahm Yes, I do have this :)

PeterCxy commented:
@connyduck Turns out we cannot implement webpush decryption support right now as Mastodon is using an old version of the webpush library that does not include a needed nonce in the notification payload. So we'll have to settle on using UnifiedPush purely as a trigger to pull notifications right now.

PeterCxy commented:
Two issues remaining for this PR:

1. The migration flow for existing users (to change the OAuth scope)
2. Updating subscription settings on the Mastodon server side according to local notification settings (i.e. whether each channel is enabled / disabled)

PeterCxy commented:
@connyduck For migration, I just added a prompt for now to let user know that they need to re-login to use push notifications. I'll definitely make dismissing that prompt permanent. The next question is how we would like to actually perform the migration -- do we just let the user re-login on their own, or do we add a new option (in Settings, for example, and direct users there in the prompt) specifically for this purpose?

connyduck commented:
> @connyduck For migration, I just added a prompt for now to let user know that they need to re-login to use push notifications. I'll definitely make dismissing that prompt permanent. The next question is how we would like to actually perform the migration -- do we just let the user re-login on their own, or do we add a new option (in Settings, for example, and direct users there in the prompt) specifically for this purpose?

A one time prompt that explains users they should relogin without logging out (or their draft & cache will be gone) should be enough. If you want to be fancy, add an special option on top of account preferences that opens the `LoginActivity` with the instance prefilled. Hide the option from migrated users.

PeterCxy commented:
@connyduck This PR is now ready for review / merge

connyduck commented on app/src/main/java/com/keylesspalace/tusky/components/login/LoginActivity.kt:
```diff
@@ -228,7 +233,15 @@ class LoginActivity : BaseActivity(), Injectable {
             { accessToken ->
-                accountManager.addAccount(accessToken.accessToken, domain)
+                if (intent.getIntExtra(LOGIN_MODE, MODE_DEFAULT) == MODE_MIGRATION) {
+                    accountManager.activeAccount?.let {
+                        it.accessToken = accessToken.accessToken
+                        it.oauthScopes = OAUTH_SCOPES
+                        accountManager.saveAccount(it)
```
We cannot be sure the user logged in with the same account, they could have mulltiple accounts on the same instance so this is dangerous. Just using `accountManager.addAccount` should work though? `updateActiveAccount` will fix it if the account already exists

connyduck commented on app/src/main/java/com/keylesspalace/tusky/receiver/NotificationBlockStateBroadcastReceiver.kt:
```diff
@@ -0,0 +1,61 @@
+    @Inject
+    lateinit var accountManager: AccountManager
+
+    override fun onReceive(context: Context?, intent: Intent?) {
```
the params can both be non-null and you save the !!s below

connyduck commented on app/src/main/java/com/keylesspalace/tusky/components/notifications/PushNotificationHelper.kt:
```diff
@@ -0,0 +1,211 @@
+import org.unifiedpush.android.connector.UnifiedPush
+import retrofit2.HttpException
+
+object PushNotificationHelper {
```
If it does not hold any state, there is no reason to create an object in Kotlin. Just make everything top level and annotate the file with `@file:JvmName("PushNotificationHelper")` to make it nice to call from Java.

connyduck commented on app/src/main/java/com/keylesspalace/tusky/components/notifications/PushNotificationHelper.kt:
```diff
@@ -0,0 +1,211 @@
+            if (currentAccountNeedsMigration(accountManager)) {
+                setMessage(R.string.dialog_push_notification_migration)
+                setPositiveButton(R.string.action_view_account_preferences) { _, _ ->
+                    PreferencesActivity.newIntent(context, PreferencesActivity.ACCOUNT_PREFERENCES).let {
```
Why not send the user straight to login here? 🤔

connyduck commented on app/src/main/java/com/keylesspalace/tusky/receiver/UnifiedPushBroadcastReceiver.kt:
```diff
@@ -0,0 +1,72 @@
+        AndroidInjection.inject(this, context)
+        Log.d(TAG, "New message for account $instance: $message")
+        val workManager = WorkManager.getInstance(context)
+        val request = OneTimeWorkRequest.from(com.keylesspalace.tusky.components.notifications.NotificationWorker::class.java)
```
import NotificationWorker?

connyduck commented on app/src/main/java/com/keylesspalace/tusky/components/login/LoginActivity.kt:
```diff
@@ -260,19 +273,28 @@ class LoginActivity : BaseActivity(), Injectable {
+
         @JvmStatic
-        fun getIntent(context: Context, mode: Boolean): Intent {
+        fun getIntent(context: Context, mode: Int): Intent {
```
default mode to MODE_DEFAULT so not all callsites have to be changed?

connyduck commented on app/src/main/java/com/keylesspalace/tusky/entity/NotificationSubscribeResult.kt:
```diff
@@ -0,0 +1,24 @@
+/* Copyright 2022 Conny Duck
```
```suggestion
/* Copyright 2022 Tusky Contributors
```

(or you name)

connyduck commented on app/src/main/res/drawable/ic_logout_24dp.xml:
```diff
@@ -0,0 +1,10 @@
+  <path
+      android:fillColor="@android:color/white"
+      android:pathData="M10.09,15.59L11.5,17l5,-5 -5,-5 -1.41,1.41L12.67,11H3v2h9.67l-2.58,2.59zM19,3H5c-1.11,0 -2,0.9 -2,2v4h2V5h14v14H5v-4H3v4c0,1.1 0.89,2 2,2h14c1.1,0 2,-0.9 2,-2V5c0,-1.1 -0.9,-2 -2,-2z"/>
+</vector>
```
We already have the same drawable, can't we reuse it?

connyduck reviewed (changes requested):
Very nice job!

Which provider did you use to test this? I tried it with the [UP-FCM Distributor](https://unifiedpush.org/users/distributors/fcm/) and it was not as fast or reliable as I hoped.

As mentioned above, please don't add fresh RxJava code, we are trying to get rid of it entirely (if possible, I'd still merge this with RxJava)

PeterCxy commented:
@connyduck These days most people use the Ntfy distributor for UP. The FCM distributor could work but I have not tested it and nor do I plan to integrate it into this PR.

PeterCxy commented:
@connyduck I believe I have addressed most of the concerns in the previous review. Before this PR is merged, I will rebase and squash all the commits into one.

connyduck commented:
Looks pretty ok - I'd be very intereseted in @charlag's opinion as well

charlag commented on app/src/main/java/com/keylesspalace/tusky/components/notifications/PushNotificationHelper.kt:
```diff
@@ -0,0 +1,216 @@
+private const val KEY_MIGRATION_NOTICE_DISMISSED = "migration_notice_dismissed"
+
+private fun anyAccountNeedsMigration(accountManager: AccountManager): Boolean =
+    accountManager.accounts.map(::accountNeedsMigration).any { it }
```
why `.map(...).any()` instead of just `.any()`

charlag commented on app/src/main/java/com/keylesspalace/tusky/components/notifications/PushNotificationHelper.kt:
```diff
@@ -0,0 +1,216 @@
+}
+
+private fun buildSubscriptionData(context: Context, account: AccountEntity): Map<String, Boolean> =
+    HashMap<String, Boolean>().apply {
```
```suggestion
    buildMap {
```

charlag commented on app/src/main/java/com/keylesspalace/tusky/components/notifications/PushNotificationHelper.kt:
```diff
@@ -0,0 +1,216 @@
+            .onSuccess {
+                Log.d(TAG, "UnifiedPush unregistration succeeded for account " + account.id)
+                // Clear the URL in database
+                account.unifiedPushUrl = ""
```
do we need to clear other parts?

PeterCxy replied:
We don't really need to -- but just to be safe I changed it to clear everything

charlag commented on app/src/main/java/com/keylesspalace/tusky/receiver/NotificationBlockStateBroadcastReceiver.kt:
```diff
@@ -0,0 +1,67 @@
+        accountManager.getAccountByIdentifier(gid)?.let { account ->
+            if (isUnifiedPushNotificationEnabledForAccount(account)) {
+                // Update UnifiedPush notification subscription
+                GlobalScope.launch { updateUnifiedPushSubscription(context, mastodonApi, accountManager, account) }
```
will the process live long enough for this? Do we need to hold something?

charlag commented on app/src/main/java/com/keylesspalace/tusky/receiver/UnifiedPushBroadcastReceiver.kt:
```diff
@@ -0,0 +1,80 @@
+        Log.d(TAG, "Endpoint available for account $instance: $endpoint")
+        accountManager.getAccountById(instance.toLong())?.let {
+            // Launch the coroutine in global scope -- it is short and we don't want to lose the registration event
+            // and there is no saner way to use structured concurrency in a receiver
```
we could probably just block here?

charlag reviewed (approved):
Looks interesting!
I'm not super happy that we have to include BC tho. Do we really need it? Is it for older versions? I would imagine it should be possible to generate EC key with plain Java (I don't remember which EC keys are compatible with what)

PeterCxy commented:
@charlag Regarding the GlobalScope usage in BroadcastReceiver: the process should be alive when it receives the event because push registration happens when the app UI is launched, and it happens quick enough so that I don't think the process being killed is a problem. However, blocking would result in a noticeable lag in the UI because the receiver runs on the UI thread. But even if the process is killed (because the user somehow closed the UI quick enough), it still does not really matter as the registration will be triggered again at the next launch anyway (we only write to the database when we are sure the notifications are set up properly).

Regarding BouncyCastle, unfortunately I'm not very familiar with the Java crypto API and have no idea how to implement this with plain JDK crypto.

@connyduck I've rebased my changes again to the tip of develop, but commit b4eda5ea65e71419233b6a350ade04baf4c95e1f introduced another db version 35. Since it is not released yet, I just merged both together. However that commit did not generate a JSON schema file -- I included that in mine.

charlag commented:
@PeterCxy I didn't realize that the event is received together with the app launch, I have no concerns then.

About key generation: I quickly searched and found this example (and others):

https://github.com/JZ-Darkal/AndroidHttpCapture/blob/9afc8805e8210df5d72e9718a745d802c59126f0/app/src/main/java/net/lightbody/bmp/mitm/keys/ECKeyGenerator.java#L37-L49

I think we just need to check that it works on our minimum version. I don't want to be the pain in the butt here but I think avoiding including something like BC is worth it.

charlag commented:
Related example from Android docs:
https://developer.android.com/reference/android/security/keystore/KeyGenParameterSpec#example:-nist-p-256-ec-key-pair-for-signingverification-using-ecdsa

PeterCxy commented:
@charlag I'd prefer not to use the Android keystore one as it can make backup & restore very painful. However I'll check if the plain JDK implementation exists on the current minSDK.

charlag commented:
@PeterCxy you don't need (and shouldn't really) use keychain for this, the relevant part of the interface is the same, all I wanted to show.

PeterCxy commented:
@charlag I've tried using the plain Java JCE APIs without BC but unfortunately there are a few obstacles. The first one is JCE is missing the same encoded form for the plain ECPoints which is needed by WebPush. This can be fixed by simply re-implementing it, which is fine. However, the Java-generated prime256v1 (aka secp256r1 in JCE) keys seem to sometimes be longer than 32 bytes, and that is incompatible with WebPush. I'm not sure why this happens yet...

PeterCxy commented:
@charlag Turns out WebPush requires the public key to be encoded in the X9.62 uncompressed point form, which is unimplemented in plain Java JCE (nor is it implemented in Android keystore). What's worse is that we cannot seem to even control whether the ECPoint objects are represented in their compressed forms or not. To avoid BouncyCastle, we would have to re-implement both X9.62 and the conversion between compressed / uncompressed representation by hand, which I think is a worse idea than including BouncyCastle.

charlag commented:
@PeterCxy thanks for checking it. I have a feeling I run into it sometime before. I think it's fine to use BC then, at least there's a reason to do that

PeterCxy commented:
@charlag Thank you for your review.

@connyduck I believe all of the concerns in the reviews have been addressed. Do you think this is ready for merge now? If so I will just go ahead and squash everything for you to merge.

PeterCxy commented:
@connyduck Rebased and squashed.

PeterCxy commented:
@connyduck Wait actually, we need a database version 36 for this as I just realized you made a new release before this. I'll do this real quick (BTW: the database schema version 35 still seems to be missing from the develop branch?)

PeterCxy commented:
Fixed. Now using database version 36.

connyduck commented:
@PeterCxy thank you! I added 35.json to develop.
Yes, using db version 36 is correct, even if there is no release in between you can never add to prevoius migrations because each commit to develop is published as nightly release.
